kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/SchemaPrinter.kt

```diff
@@ -0,0 +1,290 @@
+     * ascending by their name and appear in order of their corresponding spec section, i.e.
+     *
+     *  - 3.3 Schema
+     *  - 3.4 Types
+     *    - Scalars
+     *    - Objects
+     *    - Interfaces
+     *    - Unions
+     *    - Enums
+     *    - Input Types
+     *  - 3.13 Directives
+     *
+     * If descriptions are included, they will use single quotes, broken up on newlines, e.g.
+     *
+     * ```
+     *     "This is a long comment,"
+     *     "spanning over multiple"
+     *     "lines."
+     *     scalar DescriptedScalar
+     * ```
+     */
+    fun print(schema: __Schema): String {
+        // All actual (non-introspection) types of the schema
+        val schemaTypes = schema.types.filterNot { it.isIntrospectionType() || it.name == null }.sortedByName()
+
+        // Schema
+        //   https://spec.graphql.org/draft/#sec-Root-Operation-Types.Default-Root-Operation-Type-Names
+        //   "The type system definition language can omit the schema definition when each root operation type uses its respective default root type name and no other type uses any default root type name."
+        //   "Likewise, when representing a GraphQL schema using the type system definition language, a schema definition should be omitted if each root operation type uses its respective default root type name and no other type uses any default root type name."
+        val schemaDefinition = if (includeSchemaDefinition(schema)) {
+            buildString {
+                appendLine("schema {")
+                val indentation = "  "
+                // The query root operation type must always be provided
+                appendDescription(schema.queryType, indentation)
+                appendLine("${indentation}query: ${schema.queryType.name}")
+                // The mutation root operation type is optional; if it is not provided, the service does not support mutations
+                schema.mutationType?.let {
+                    appendDescription(it, indentation)
+                    appendLine("${indentation}mutation: ${it.name}")
+                }
+                // Similarly, the subscription root operation type is also optional; if it is not provided, the service does not support subscriptions
+                schema.subscriptionType?.let {
+                    appendDescription(it, indentation)
+                    appendLine("${indentation}subscription: ${it.name}")
+                }
+                appendLine("}")
+            }
+        } else {
+            ""
+        }
+
+        // Scalars
+        //   https://spec.graphql.org/draft/#sec-Scalars.Built-in-Scalars
+        //   "When representing a GraphQL schema using the type system definition language, all built-in scalars must be omitted for brevity."
+        val scalars = buildString {
+            schemaTypes.filter { it.kind == TypeKind.SCALAR && !it.isBuiltInScalar() }.forEachIndexed { index, type ->
+                if (index > 0) {
```

Comment:
Question:

Is this empty line really needed for it to be valid SDL, or is this just for pretty printing?

The same question goes for all such implementations in here

Reply:
I couldn't find much actual spec regarding SDL. Pretty sure that empty lines are not required but IMHO readability for humans is one of the core features for SDL, so I'd like to keep it.